bluemira/equilibria/fem_fixed_boundary/fem_magnetostatic_2D.py

```diff
@@ -183,3 +196,232 @@ def calculate_b(self):
+        self._plot_array(
+            points, curr_data, "J current at iteration", i_iter, contour=False
+        )
+
+    def _plot_array(self, points, array, title, i_iter, contour=True):
+        axis, cntr, _ = plot_scalar_field(
+            points[:, 0],
+            points[:, 1],
+            array,
+            levels=20,
+            axis=None,
+            tofill=True,
+            contour=contour,
+        )
+        plt.title(f"{title} {i_iter}")
+        plt.show()
+
+    def solve(
+        self,
+        pprime,
+        ffprime,
+        curr_target,
+        dirichlet_bc_function=None,
+        dirichlet_marker=None,
+        neumann_bc_function=None,
+        tol=1e-5,
+        max_iter=10,
+        relaxation=1,
+        verbose=False,
+        verbose_plot=False,
+    ):
+        """Solves the GS problem given pprime and ffprime"""
+        points = self.mesh.coordinates()
+
+        if verbose_plot:
+            self.g_func = self._create_g_func(pprime, ffprime, curr_target)
+            self._plot_current_iteration(points, 0)
+
+        self.g = self._create_g(pprime, ffprime, curr_target)
+
+        super().solve(
+            self.g, dirichlet_bc_function, dirichlet_marker, neumann_bc_function
+        )
+        self._update_curr(curr_target)
+
+        eps = 1.0
+        i = 0  # iteration counter
+        while eps > tol and i < max_iter:
+            i += 1
+
+            prev_psi = self.psi.vector()[:]
+
+            if verbose_plot:
+                self._plot_current_iteration(points, i)
+
+            prev = np.array([self.psi_norm_2d(p) for p in points])
+
+            if verbose_plot:
+                self._plot_array(prev, "Normalized magnetic coordinate at iteration", i)
+
+            # print(f"Magnetic coordinate range = [{np.min(prev)}:{np.max(prev)}]")
+
+            # prev = np.array([self.psi(p) for p in points])
+            # prev = self.psi.compute_vertex_values()
+            # curr_tot = dolfin.assemble(self.g * dx())
+            # print(f"pre - curr_tot = {curr_tot} - curr_dens = {self.g}")
+
+            super().solve(
+                self.g, dirichlet_bc_function, dirichlet_marker, neumann_bc_function
+            )
+
+            new = np.array([self.psi_norm_2d(p) for p in points])
+            diff = new - prev
+
+            if verbose_plot:
+                self._plot_array(diff, "GS error at iteration", i)
+
+            # diff = self.psi.compute_vertex_values() - prev
+            eps = np.linalg.norm(diff, ord=2) / np.linalg.norm(new, ord=2)
+
+            # curr_tot = dolfin.assemble(self.g * dx())
+            # print(f"post - curr_tot = {curr_tot} - curr_dens = {self.g}")
+            bluemira_print_flush(
+                f"iter = {i} eps = {eps:.3E} psi_ax : {self.psi_ax:.2f}"
+            )
+
+            new_psi = dolfin.Function(self.V)
+            new_psi.set_allow_extrapolation(True)
+
+            self.psi.vector()[:] = (1 - relaxation) * self.psi.vector()[
```

Comment:
OK, so I understand that dolfin sometimes works in mysterious ways... It's hard for me to read this "Python" and understand that this is updating the psi vector in place! I assume that is what it is supposed to be doing.